File: apps/ios/Packages/StoryArcKit/Sources/LibraryFeature/KavitaSheet.swift
public import SwiftUI

internal import DesignSystem
internal import Kavita
public import StoryArcCore

/// Adding a Kavita server.
///
/// The same shape as ``CatalogueSheet``, because it is the same job: an address, whatever
/// proof of identity the server wants, and a confirmation before anything is saved.
public struct KavitaSheet: View {
    @Environment(\.theme) private var theme
    @Environment(\.dismiss) private var dismiss

    private let connection: KavitaConnection
    private let onAdd: (Source) -> Void

    public init(connection: KavitaConnection, onAdd: @escaping (Source) -> Void) {
        self.connection = connection
        self.onAdd = onAdd
    }

    public var body: some View {
        NavigationStack {
            ScrollView {
                VStack(alignment: .leading, spacing: StoryArcSpace.lg) {
                    address
                    key

                    Button {
                        Task { await connection.connect() }
                    } label: {
                        Text("kavita.connect", bundle: .module)
                            .frame(maxWidth: .infinity)
                            .padding(.vertical, StoryArcSpace.xs)
                    }
                    // **The one thing on this form that does something, shaped like it.**
                    // `.borderedProminent` renders as a plain grey capsule while it is
                    // disabled — which is the state a reader meets it in, because the form
                    // starts empty — the same colour, height and corner as the field above
                    // it. The 2026-09-02 sweep called it "a fifth thing to type into", and
                    // the frames agree.
                    //
                    // `.glassProminent` is how this app emphasises: `design.md` §5 keeps
                    // chrome glass untinted so it picks up what is behind it, and the
                    // prominent variant is the one meant to carry a tint — a `.tint` on plain
                    // `.glass` tints the *material* and flattens it, which
                    // `GlassIsUntintedTests` fails the build over. `DetailActions` already
                    // uses exactly this trio for the *Read* button, which is the same job:
                    // the one functional action on the screen.
                    .buttonStyle(.glassProminent)
                    .controlSize(.large)
                    .tint(theme.accent)
                    .disabled(!connection.canConnect)

                    switch connection.step {
                    case .entering:
                        EmptyView()
                    case .connecting:
                        Label {
                            Text("catalogue.connecting", bundle: .module)
                        } icon: {
                            ProgressView()
                        }
                        .textRole(.subheadline)
                        .foregroundStyle(theme.palette.textSecondary)
                    case let .confirmed(identity):
                        confirmation(identity)
                    case let .failed(message):
                        CatalogueFailure(message: message) {
                            Task { await connection.connect() }
                        }
                    }
                }
                .padding(StoryArcSpace.gutter)
            }
            .background(theme.palette.surfaceCanvas)
            .navigationTitle(Text("kavita.title", bundle: .module))
            .toolbar {
                ToolbarItem(placement: .cancellationAction) {
                    Button(role: .cancel) { dismiss() }
                }
            }
        }
    }

    @ViewBuilder
    private var address: some View {
        VStack(alignment: .leading, spacing: StoryArcSpace.sm) {
            Text("kavita.address.label", bundle: .module)
                .textRole(.headline)
                .foregroundStyle(theme.palette.textPrimary)

            TextField(
                text: Binding(get: { connection.address }, set: { connection.address = $0 }),
                prompt: Text("kavita.address.prompt", bundle: .module)
            ) {
                Text("kavita.address.label", bundle: .module)
            }
            .labelsHidden()
            // **The field was nameless to a screen reader, on the one form where a wrong
            // character means a refused connection.** Measured by an XCUITest dump on
            // 2026-09-11: textFields: [""] and secureTextFields: [""]. The name is drawn as a
            // separate headline above the field and hidden on the field itself, so VoiceOver
            // said "text field" and nothing else. From the same key the headline draws, so
            // the two cannot drift.
            .accessibilityLabel(Text("kavita.address.label", bundle: .module))
            .accessibilityHint(Text("kavita.address.hint", bundle: .module))
            .textFieldStyle(.roundedBorder)
            .autocorrectionDisabled()
            #if os(iOS)
            .textInputAutocapitalization(.never)
            .keyboardType(.URL)
            #endif

            Text("kavita.address.hint", bundle: .module)
                .textRole(.footnote)
                .foregroundStyle(theme.palette.textSecondary)
                // Spoken once, as the field's hint. It stays on screen for everyone else;
                // read as its own element it would be the same sentence twice.
                .accessibilityHidden(true)
        }
    }

    /// The key field, hidden when the address already carries one.
    ///
    /// Asking for something the reader has already given is how a form makes someone feel
    /// they typed it wrong.
    @ViewBuilder
    private var key: some View {
        if connection.addressCarriesKey {
            Label {
                Text("kavita.keyFromAddress", bundle: .module)
                    .textRole(.footnote)
            } icon: {
                Image(systemName: "checkmark.circle.fill")
                    .foregroundStyle(StoryArcColor.Status.success)
            }
            .foregroundStyle(theme.palette.textSecondary)
        } else {
            VStack(alignment: .leading, spacing: StoryArcSpace.sm) {
                Text("kavita.key.label", bundle: .module)
                    .textRole(.headline)
                    .foregroundStyle(theme.palette.textPrimary)

                SecureField(
                    text: Binding(get: { connection.apiKey }, set: { connection.apiKey = $0 }),
                    prompt: Text("kavita.key.label", bundle: .module)
                ) {
                    Text("kavita.key.label", bundle: .module)
                }
                .labelsHidden()
                // The secret field, and the one this defect mattered most on. See the address
                // field above for what was measured.
                .accessibilityLabel(Text("kavita.key.label", bundle: .module))
                .accessibilityHint(Text("kavita.key.hint", bundle: .module))
                .textFieldStyle(.roundedBorder)

                Text("kavita.key.hint", bundle: .module)
                    .textRole(.footnote)
                    .foregroundStyle(theme.palette.textSecondary)
                    .accessibilityHidden(true)
            }
        }
    }

    @ViewBuilder
    private func confirmation(_ identity: KavitaIdentity) -> some View {
        VStack(alignment: .leading, spacing: StoryArcSpace.md) {
            Label {
                Text("kavita.confirmed \(identity.username)", bundle: .module)
                .textRole(.headline)
                .foregroundStyle(theme.palette.textPrimary)
            } icon: {
                Image(systemName: "checkmark.circle.fill")
                    .foregroundStyle(StoryArcColor.Status.success)
            }

            Button {
                guard let source = connection.source() else { return }
                onAdd(source)
                dismiss()
            } label: {
                Text("kavita.add", bundle: .module)
                    .frame(maxWidth: .infinity)
                    .padding(.vertical, StoryArcSpace.xs)
            }
            .buttonStyle(.borderedProminent)
        }
    }
}

File: apps/android/feature/library/src/main/kotlin/app/storyarc/feature/library/KavitaSheet.kt
package app.storyarc.feature.library

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Button
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.input.PasswordVisualTransformation
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import app.storyarc.core.designsystem.theme.LocalStoryArcPalette
import app.storyarc.core.designsystem.tokens.StoryArcSpace
import app.storyarc.core.model.Source

/**
 * Adding a Kavita server.
 *
 * The same shape as [CatalogueSheet], because it is the same job: an address, whatever proof
 * of identity the server wants, and a confirmation before anything is saved.
 */
@OptIn(androidx.compose.material3.ExperimentalMaterial3Api::class)
@Composable
fun KavitaSheet(
    connection: KavitaConnection,
    onAdd: (Source) -> Unit,
    onDismiss: () -> Unit,
) {
    val palette = LocalStoryArcPalette.current
    val step by connection.step.collectAsStateWithLifecycle()
    val address by connection.address.collectAsStateWithLifecycle()
    val apiKey by connection.apiKey.collectAsStateWithLifecycle()
    val carriesKey = connection.carriesKey(address)

    ModalBottomSheet(onDismissRequest = onDismiss) {
        Column(
            modifier = Modifier
                .verticalScroll(rememberScrollState())
                .padding(horizontal = StoryArcSpace.gutter)
                .padding(bottom = StoryArcSpace.xl),
            verticalArrangement = Arrangement.spacedBy(StoryArcSpace.md),
        ) {
            Text(
                text = stringResource(R.string.kavita_title),
                style = MaterialTheme.typography.headlineSmall,
                color = palette.textPrimary,
            )

            OutlinedTextField(
                value = address,
                onValueChange = { connection.address.value = it },
                label = { Text(stringResource(R.string.kavita_address_label)) },
                placeholder = { Text(stringResource(R.string.kavita_address_prompt)) },
                singleLine = true,
                keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Uri),
                modifier = Modifier.fillMaxWidth(),
            )
            Text(
                text = stringResource(R.string.kavita_address_hint),
                style = MaterialTheme.typography.bodySmall,
                color = palette.textSecondary,
            )

            // Hidden when the address already carries a key: asking for something the reader
            // has already given is how a form makes someone feel they typed it wrong.
            if (carriesKey) {
                Text(
                    text = stringResource(R.string.kavita_key_from_address),
                    style = MaterialTheme.typography.bodySmall,
                    color = palette.accent,
                )
            } else {
                OutlinedTextField(
                    value = apiKey,
                    onValueChange = { connection.apiKey.value = it },
                    label = { Text(stringResource(R.string.kavita_key_label)) },
                    singleLine = true,
                    visualTransformation = PasswordVisualTransformation(),
                    modifier = Modifier.fillMaxWidth(),
                )
                Text(
                    text = stringResource(R.string.kavita_key_hint),
                    style = MaterialTheme.typography.bodySmall,
                    color = palette.textSecondary,
                )
            }

            Button(
                onClick = { connection.connect() },
                enabled = address.isNotBlank() &&
                    (carriesKey || apiKey.isNotBlank()) &&
                    step !is KavitaConnection.Step.Connecting,
                modifier = Modifier.fillMaxWidth(),
            ) {
                Text(stringResource(R.string.kavita_connect))
            }

            when (val current = step) {
                is KavitaConnection.Step.Entering -> Unit

                is KavitaConnection.Step.Connecting ->
                    CircularProgressIndicator(strokeWidth = 2.dp, modifier = Modifier.padding(2.dp))

                is KavitaConnection.Step.Confirmed -> {
                    Text(
                        text = stringResource(
                            R.string.kavita_confirmed,
                            current.identity.username,
                            current.identity.version.toString(),
                        ),
                        style = MaterialTheme.typography.titleMedium,
                        color = palette.textPrimary,
                    )
                    Button(
                        onClick = {
                            connection.source()?.let(onAdd)
                            onDismiss()
                        },
                        modifier = Modifier.fillMaxWidth(),
                    ) {
                        Text(stringResource(R.string.kavita_add))
                    }
                }

                is KavitaConnection.Step.Failed -> {
                    Text(
                        text = current.message,
                        style = MaterialTheme.typography.bodyMedium,
                        color = palette.textPrimary,
                    )
                    OutlinedButton(onClick = { connection.connect() }) {
                        Text(stringResource(R.string.catalogue_retry))
                    }
                }
            }
        }
    }
}
